modules/externalipam/ip_cache.rb

```diff
@@ -0,0 +1,133 @@
+require 'yaml'
+require 'json'
+require 'monitor'
+require 'concurrent'
+require 'time'
+require 'externalipam/ipam_helper'
+require 'singleton'
+
+module Proxy::Ipam
+  # Class for managing temp in-memory cache to prevent same IP's being suggested in race conditions
+  class IpCache
+    include Singleton
+    include Proxy::Log
+    include Proxy::Ipam::IpamHelper
+
+    DEFAULT_CLEANUP_INTERVAL = 60
+
+    def initialize
+      @m = Monitor.new
+      init_cache
+      start_cleanup_task
+    end
+
+    def provider_name(provider)
+      @provider = provider
+    end
+
+    def set_group(group, value)
+      @ip_cache[group.to_sym] = value
+    end
+
+    def get_group(group)
+      @ip_cache[group.to_sym]
+    end
+
+    def get_cidr(group, cidr)
+      @ip_cache[group.to_sym][cidr.to_sym]
+    end
+
+    def get_ip(group_name, cidr, mac)
+      @ip_cache[group_name.to_sym][cidr.to_sym][mac.to_sym][:ip]
```

Comment:
Wouldn't this be safer?
```suggestion
      @ip_cache.dig(group_name.to_sym, cidr.to_sym, mac.to_sym, :ip)
```

If for some reason the group (or cidr or mac) is not found, your code will result in `nil[]` which raises an exception where `dig()` will return nil.